File: csharp/02_integrate_agent_with_llm/Agent/Agent.Application/Agent.cs
namespace Agent.Application;

public class Agent()
{
    private readonly IUserInput _input;
    private readonly IDisplay _display;
    private readonly ILanguageModel _model;

    public Agent(IUserInput input, ILanguageModel model, IDisplay display) : this()
    {
        _input = input;
        _display = display;
        _model = model;
    }

    public void Run()
    {
        var userInput = _input.GetInput();
        _display.Show("User: " + userInput);
        var message = _model.Prompt([new Message("user", userInput)]);
        _display.Show("Assistant: " + message.Content);
    }
}

File: java/template/agent-application/src/main/java/com/workshop/agent/application/Agent.java
package com.workshop.agent.application;

import java.util.ArrayList;
import java.util.List;

public class Agent {
    private final String systemPrompt;
    private final UserInput input;
    private final Display display;
    private final LanguageModel model;
    private final Tool tool;

    public Agent(String systemPrompt, UserInput input, LanguageModel model, Tool tool, Display display) {
        this.systemPrompt = systemPrompt;
        this.input = input;
        this.display = display;
        this.model = model;
        this.tool = tool;
    }

    public void run() {
        List<Message> context = new ArrayList<>();
        context.add(new Message("system", systemPrompt));

        while (true) {
            String userInput = input.getInput();
            if (userInput == null || userInput.isBlank()) {
                break;
            }
            Message userMessage = new Message("user", userInput);
            display.show(userMessage);
            context.add(userMessage);

            Message answer = model.prompt(context);
            display.show(answer);
            context.add(answer);

            while (true) {
                var toolResult = tool.parseAndExecute(answer.content());
                if (toolResult.isEmpty()) {
                    break;
                }

                Message toolResultMessage = new Message("user", toolResult.get());
                context.add(toolResultMessage);
                display.show(toolResultMessage);
                answer = model.prompt(context);
                display.show(answer);
                context.add(answer);
            }
        }
    }
}
